### DISCUSSION TITLE
New onQuery handler?

### DISCUSSION THREAD

**Discussion by denisborovikov** [TanStack/query#278] [General] [answered]
In some cases, we might want to optimistically update the cache before or at the same time, when the mutation occurs. 

We can do it by using `queryCache.setQueryData` and then calling mutation function  like [here](https://github.com/tannerlinsley/react-query/blob/master/examples/optimistic-updates/pages/index.js#L32) 

```js
queryCache.setQueryData('todos', [...data, text])
...
mutatePostTodo(text)
```

But as I can see, we can't encapsulate an optimistic cache update into the custom hook. If we would have a handler, that executes at the same time as the mutation, we could do it:

```js
function useCustomDelete() {
  return useMutation(foo, { onQuery: variables => queryCache.setQueryData(...)})
}

const [delete] = useCustomDelete();
...
delete(variables);
```

What do you think?

**Comment by tannerlinsley** [comment]
Ooooooo very interesting. How about onMutate?

**Reply by denisborovikov** [reply]
Yes, it makes more sense in this context.

**Comment by tannerlinsley** [comment]
Would it be worth discussing how a general rollback would take place too? Eg if the mutation falls?

**Reply by denisborovikov** [reply]
I think, existing onError/onSettled should perfectly work here.

**Reply by tannerlinsley** [reply]
What about storing the previous value before the optimistic update?

**Reply by denisborovikov** [reply]
I see what you mean. I didn't think about it in this way. I meant we can refetch query from the server in case of error.

**Reply by denisborovikov** [reply]
I think this will work. 

```js
function useCustomDelete() {
  const rollback = queryCache. getQueryData(...);
  return useMutation(foo, { 
    onMutation: variables => queryCache.setQueryData(...)},
    onError: () => queryCache.setQueryData(rollback),
  )
}

const [delete] = useCustomDelete();
...
delete(variables);
```

**Comment by tannerlinsley** [comment]
What do you think about this?

# Automatic Rollback for Failed Mutations

When you optimistically update your state before performing a mutation, there is a non-zero chance that the mutation will fail. In that event, you would normally refetch your queries that you optimistically updated. In some circumstances though, refetching may not work correctly and the mutation error could represent some type of server issue that won't make it possible to refetch. In this event, you can instead choose to rollback your update.

To do this, `useMutation` provides an `onMutate` handler that can achieve both the optimistic update and the rollback logic in one single API. First, you define your optimistic update, then return a function to run in the event that the mutation fails:

```js
useMutation(updateTodo, {
  onMutate: (newTodo) => {
    // Store the previous value in this closure
    const previousTodo = queryCache.getQueryData(['todos', newTodo.id])

    // Optimistically update the query data
    queryCache.setQueryData(['todos', newTodo.id], newTodo)

    // Return a rollback function
    return (err) => {
      queryCache.setQueryData(['todos', newTodo.id], previousTodo)
    }
  }
})
```

**Reply by tannerlinsley** [reply]
@kentcdodds, @cherniavskii, What do you think of this?

**Reply by denisborovikov** [reply]
Wow, very interesting. I was thinking about returning a rollback value from onMutate and pass it as an argument to onError, your option is much more elegant and straightforward.

**Reply by tannerlinsley** [reply]
Well, actually, that's not a bad idea either. This approach would mean there are 2 error handlers technically, with your idea, just one.

**Reply by tannerlinsley** [reply]
So it would look more like this: 

```js
useMutation(updateTodo, {
  // When mutate is called:
  onMutate: (newTodo) => {
    // Store the previous value in this closure
    const previousTodo = queryCache.getQueryData(['todos', newTodo.id])

    // Optimistically update the query data
    queryCache.setQueryData(['todos', newTodo.id], newTodo)

    // Return a rollback value
    return previousTodo
  },
  // If the mutation fails:
  onError: (err, newTodo, previousTodo) => {
    queryCache.setQueryData(['todos', newTodo.id], previousTodo)
  }
  // If the mutation was successful:
  onSuccess: newTodo => queryCache.setQueryData(['todos', newTodo.id], newTodo)
})
```

**Reply by tannerlinsley** [reply]
While the first approach is very useEffect-esque with the rollback function, I like the idea of only having one single handler for each side-effect

**Reply by denisborovikov** [reply]
My concern is that originally I thought about a general-purpose handler which fires at the same time as mutation happens. The second thought was that we can use it for optimistic updates. But it's not limited to this use-case, so returning value or function may cause questions. Maybe it makes sense to name the handler specifically for optimistic updates.

**Reply by tannerlinsley** [reply]
I think it's totally fine to define the API up front that whatever you return will be available in the onError handler. As long as it's documented well, I think `onMutate` is a fine name, mostly because it corresponds to *what* is happening and *when*, instead of assuming a specific purpose like `onMutateRollback`, which is really just one interpretation of how to use the side-effect.

**Reply by cherniavskii** [reply]
I like "return value" more, because there's less magic in it
Also, `onMutate` naming seems good to me

**Comment by cherniavskii** [comment]
> But as I can see, we can't encapsulate an optimistic cache update into the custom hook.

Did you consider something like this?

```js
function useCustomDelete() {
  const [mutate, ...mutation] = useMutation(foo);

  function customMutate(...args) {
    queryCache.setQueryData(...)
    return mutate(...args)
  }

  return [customMutate, ...mutation];
}

const [delete] = useCustomDelete();
...
delete(variables);
```

**Reply by tannerlinsley** [reply]
You would still end up needing to get the previous value, then extend the functionality of the onError handler to revert back. It's not that it's not possible to do right now, and more that the API doesn't afford to it.

**Reply by denisborovikov** [reply]
Yes, this is also possible but looks hacky for me. You'll probably also need to use `useMemo` to preserve reference equality what makes the code even more verbose. The logic base on handlers seems more idiomatic.

**Reply by tannerlinsley** [reply]
Agreed. I think [this](https://github.com/tannerlinsley/react-query/discussions/278#discussioncomment-1885) approach is my favorite right now.

**Reply by cherniavskii** [reply]
Just wanted to make sure it's possible right now

**Comment by tannerlinsley** [comment, marked as answer]
Here are the final docs:

# Automatic Rollback for Failed Mutations

When you optimistically update your state before performing a mutation, there is a non-zero chance that the mutation will fail. In most cases, you can just trigger a refetch for your optimistic queries to revert them to their true server state. In some circumstances though, refetching may not work correctly and the mutation error could represent some type of server issue that won't make it possible to refetch. In this event, you can instead choose to rollback your update.

To do this, `useMutation`'s `onMutate` handler option allows you to return a snapshotted rollback value that will later be passed to both `onError` and `onSettled` handlers as the last argument:

```js
useMutation(updateTodo, {
  // When mutate is called:
  onMutate: (newTodo) => {
    // Snapshot the previous value
    const previousTodo = queryCache.getQueryData(['todos', newTodo.id])

    // Optimistically update to the new value
    queryCache.setQueryData(['todos', newTodo.id], newTodo)

    // Return the snapshotted value
    return previousTodo
  },
  // If the mutation fails, use the value returned from onMutate to roll back
  onError: (err, newTodo, previousTodo) => {
    queryCache.setQueryData(['todos', newTodo.id], previousTodo)
  }
  // If the mutation was successful:
  onSuccess: newTodo => {
    queryCache.setQueryData(['todos', newTodo.id], newTodo)
  }
})
```

You can also use the `onSettled` function in place of the separate `onError` and `onSuccess` handlers if you wish:

```js
useMutation(updateTodo, {
  // ...
  onSettled: (newTodo, error, variables, previousTodo) => {
    if (error) {
      queryCache.setQueryData(['todos', newTodo.id], previousTodo)
    } else {
      queryCache.setQueryData(['todos', newTodo.id], newTodo)
    }
  },
})
```

**Reply by cherniavskii** [reply]
```diff
- onError: (err, newTodo, previousTodo) => {
+ onError: (err, variables, previousTodo) => {
```
?

**Reply by cherniavskii** [reply]
actually it doesn't matter

**Reply by tannerlinsley** [reply]
Yeah, that's actually just for the example/walkthrough portion of the docs. The actual API docs have it listed more generically.
